core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2049,12 +2054,31 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     * res0: Unit = ()
     * }}}
     */
-  def parEvalMapUnordered[F2[x] >: F[
-    x
-  ]: Concurrent, O2](
+  def parEvalMapUnordered[F2[x] >: F[x]: Concurrent, O2](
       maxConcurrent: Int
   )(f: O => F2[O2]): Stream[F2, O2] =
-    map(o => Stream.eval(f(o))).parJoin(maxConcurrent)
+    Stream
+      .eval {
+        (
+          Semaphore[F2](maxConcurrent),
+          Queue.bounded[F2, Option[O2]](1)
+        ).tupled
+      }
+      .flatMap { case (semaphore, queue) =>
+        Stream
+          .fromQueueNoneTerminated(queue)
+          .concurrently {
+            noneTerminate
+              .flatMap { opt =>
+                Stream.exec {
+                  opt.fold(queue.offer(none)) { el =>
+                    val running = f(el).flatMap(result => queue.offer(result.some))
+                    semaphore.acquire >> running.guarantee(semaphore.release).start.void
```

Comment:
Does a `.start.void` leak fibers? Is this ignorings errors from inner streams?

We would probably need to add to this a common `interrupt` deferred, as used in other combinators, so that we can stop all inner streams once any of them has an error.

(of course this is still a draft)

Reply:
Yes, I currently ignore errors from the effects.

TBH, I don't know about fibers leakage... Can you give me some docs on the subject?
However, we do not allocate here more fibers than `maxConcurrent`, because `acquire` is executed synchronously and only after acquisition we launch the remaining IO. Moreover, no fiber should live more, than running + guarantee(I suppose :flushed:)